src/coreclr/jit/emitriscv64.cpp

```diff
@@ -1258,19 +1258,33 @@ void emitter::emitLoadImmediate(emitAttr size, regNumber reg, ssize_t imm)
     INT32 high19 = ((int32_t)(high31 + 0x800)) >> 12;
 
     emitIns_R_I(INS_lui, size, reg, high19);
-    emitIns_R_R_I(INS_addiw, size, reg, reg, high31 & 0xFFF);
+    if (high31 & 0xFFF) {
+        emitIns_R_R_I(INS_addiw, size, reg, reg, high31 & 0xFFF);
+    }
 
     // And load remaining part part by batches of 11 bits size.
     INT32 remainingShift = msb - 30;
+
+    UINT32 shiftAccumulator = 0;
     while (remainingShift > 0)
     {
         UINT32 shift = remainingShift >= 11 ? 11 : remainingShift % 11;
```

Comment:
We could utilize all 12 bits by adding the sign bit like in line 1258. Up to you if this PR should include it.

Reply:
Er, that would require a change to walk the immediate from least to most significant bits, I think we can do that later.

Reply:
I think 11 bits is enough because 12 bits cannot reduce instructions and it needs handling for sign bit.